lib/srv/db/spanner/grpcserver.go

```diff
@@ -0,0 +1,467 @@
+	if err := e.authorizeRPCRequest(ctx, req.GetSession(), "Read", req, false); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	clt, err := e.getClient(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	res, err := clt.Read(ctx, req)
+	return res, trace.Wrap(err)
+}
+
+func (e *Engine) BeginTransaction(ctx context.Context, req *spannerpb.BeginTransactionRequest) (*spannerpb.Transaction, error) {
+	if err := e.authorizeRPCRequest(ctx, req.GetSession(), "BeginTransaction", req, false); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	clt, err := e.getClient(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	res, err := clt.BeginTransaction(ctx, req)
+	return res, trace.Wrap(err)
+}
+
+func (e *Engine) Commit(ctx context.Context, req *spannerpb.CommitRequest) (*spannerpb.CommitResponse, error) {
+	if err := e.authorizeRPCRequest(ctx, req.GetSession(), "Commit", req, false); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	clt, err := e.getClient(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	res, err := clt.Commit(ctx, req)
+	return res, trace.Wrap(err)
+}
+
+func (e *Engine) Rollback(ctx context.Context, req *spannerpb.RollbackRequest) (*emptypb.Empty, error) {
+	if err := e.authorizeRPCRequest(ctx, req.GetSession(), "Rollback", req, false); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	clt, err := e.getClient(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	res, err := clt.Rollback(ctx, req)
+	return res, trace.Wrap(err)
+}
+
+func (e *Engine) PartitionRead(ctx context.Context, req *spannerpb.PartitionReadRequest) (*spannerpb.PartitionResponse, error) {
+	if err := e.authorizeRPCRequest(ctx, req.GetSession(), "PartitionRead", req, false); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	clt, err := e.getClient(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	res, err := clt.PartitionRead(ctx, req)
+	return res, trace.Wrap(err)
+}
+
+func (e *Engine) BatchWrite(req *spannerpb.BatchWriteRequest, stream spannerpb.Spanner_BatchWriteServer) error {
+	ctx := stream.Context()
+	if err := e.authorizeRPCRequest(ctx, req.GetSession(), "BatchWrite", req, false); err != nil {
+		return trace.Wrap(err)
+	}
+	clt, err := e.getClient(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	cc, err := clt.BatchWrite(ctx, req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return trace.Wrap(proxyServerStream(cc, stream))
+}
+
+func (e *Engine) ExecuteSql(ctx context.Context, req *spannerpb.ExecuteSqlRequest) (*spannerpb.ResultSet, error) {
+	skipAudit := isQuerySpam(req.GetSql())
+	if err := e.authorizeRPCRequest(ctx, req.GetSession(), "ExecuteSql", req, skipAudit); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	clt, err := e.getClient(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	res, err := clt.ExecuteSql(ctx, req)
+	return res, trace.Wrap(err)
+}
+
+func (e *Engine) PartitionQuery(ctx context.Context, req *spannerpb.PartitionQueryRequest) (*spannerpb.PartitionResponse, error) {
```

Comment:
It might be possible to make it *more* generic, but I don't think there's a way to do it generically enough to avoid having to add each method to the engine.

For example, we could do authorization in the interceptors, and use runtime assertions to extract the `targetID` and pass it to `authorizeRPCRequest`:

```go
// unaryServerLoggingInterceptor is gRPC middleware that logs some debug info.
func (e *Engine) unaryServerLoggingInterceptor() grpc.UnaryServerInterceptor {
	return func(
		ctx context.Context,
		req interface{},
		info *grpc.UnaryServerInfo,
		handler grpc.UnaryHandler,
	) (interface{}, error) {
		targetID, err := getTargetID(req)
		if err != nil {
			return nil, trace.Wrap(err)
		}
		method, ok := grpc.Method(ctx)
		if !ok {
			return nil, trace.BadParameter("unknown method")
		}
		if err := e.authorizeRPCRequest(ctx, targetID, method, req, false /*skipAudit*/); err != nil {
			return nil, trace.Wrap(err)
		}
		res, err := handler(ctx, req)
		e.Log.WithError(err).
			WithField("full_method", info.FullMethod).
			Trace("Handled unary gRPC request")
		return res, err
	}
}

func getTargetID(req any) (string, error) {
	switch idGetter := req.(type) {
	case interface{ GetName() string }:
		return idGetter.GetName(), nil
	case interface{ GetDatabase() string }:
		return idGetter.GetDatabase(), nil
	case interface{ GetSession() string }:
		return idGetter.GetSession(), nil
	}
	return "", trace.BadParameter("unable to extract target database ID")
}
```

I considered trying to make it more generic like this, but it doesn't seem to save much code and I think it's more complicated. What do you think?

Reply:
I was hoping we can write a generic forwarder here, for any possible method, but I took a closer look and that doesn't seem straightforward or a win for complexity.